sgv2-service-common/src/main/java/io/stargate/sgv2/common/grpc/DefaultStargateBridgeClient.java

```diff
@@ -107,6 +110,91 @@ public CompletionStage<Response> executeQueryAsync(Query query) {
     return observer.asFuture();
   }
 
+  @Override
+  public CompletionStage<Response> executeQueryAsync(
+      String keyspaceName,
+      String tableName,
+      Function<Optional<Schema.CqlTable>, Query> queryProducer) {
+    String decoratedKeyspaceName = addTenantPrefix(keyspaceName);
+    CqlKeyspaceDescribe keyspace = keyspaceCache.getIfPresent(decoratedKeyspaceName);
+    if (keyspace == null) {
+      // Nothing in our local cache, check if the bridge has a more recent version
+      return forceFetchAndExecute(keyspaceName, decoratedKeyspaceName, tableName, queryProducer);
+    } else {
+      // Build the query with our local version
+      Query query;
+      try {
+        query = produceQuery(queryProducer, Optional.of(keyspace), tableName);
+      } catch (Throwable t) {
+        // Always retry with the latest version, in case the error is caused by stale metadata
+        // (e.g. using a table that has not yet replicated to our local cache)
+        return forceFetchAndExecute(keyspaceName, decoratedKeyspaceName, tableName, queryProducer);
+      }
+      // Execute optimistically
+      ClientCall<QueryWithSchema, QueryWithSchemaResponse> call =
+          channel.newCall(StargateBridgeGrpc.getExecuteQueryWithSchemaMethod(), callOptions);
+      UnaryStreamObserver<QueryWithSchemaResponse> observer = new UnaryStreamObserver<>();
+      ClientCalls.asyncUnaryCall(
+          call,
+          QueryWithSchema.newBuilder()
+              .setQuery(query)
+              .setKeyspaceName(keyspaceName)
+              .setKeyspaceHash(keyspace.getHash().getValue())
+              .build(),
+          observer);
+      return observer
+          .asFuture()
+          .thenCompose(
+              withSchemaResponse -> {
+                QueryWithSchemaResponse.InnerCase innerCase = withSchemaResponse.getInnerCase();
+                switch (innerCase) {
+                  case RESPONSE:
+                    // Success, the keyspace hadn't changed on the bridge
+                    return CompletableFuture.completedFuture(withSchemaResponse.getResponse());
+                  case NEW_KEYSPACE:
+                    // The keyspace was updated, retry with that version
+                    CqlKeyspaceDescribe ks = withSchemaResponse.getNewKeyspace();
+                    keyspaceCache.put(decoratedKeyspaceName, ks);
+                    return executeQueryAsync(
```

Comment:
Is there any danger this could loop indefinitely? That is, what mechanism makes the retry called only once?

Reply:
This is not a recursive call, it invokes the original `executeQueryAsync()` method. If `produceQuery()` throws this time, the exception will propagate.

Reply:
Ahhhhh. I misread the invocation so it looked it was recursive call. But it isn't, there's just 1 arg.
Yes, that works.